contributors/design-proposals/api-machinery/customresource-conversion-webhook.md

```diff
@@ -0,0 +1,861 @@
+
+**Step 2**: Create a CR within that one version:
+
+|cr1.yaml|
+|-|
+```yaml
+apiVersion: "stable.example.com/v1"
+kind: CronTab
+metadata:
+  name: my-new-cron-object
+spec:
+  cronSpec: "* * * * */5"
+  image: my-awesome-cron-image
+```
+
+
+And create it:
+
+```bash
+Kubectl create -f cr1.yaml
+```
+
+
+**Step 3**: Decide to introduce a new version of the API.
+
+**Step 3a**: Write a new OpenAPI data model for the new version (see previous section).  Use of a data model is not required, but it is recommended.
+
+**Step 3b**: Write conversion webhook and deploy it as a service named `crontab_conversion`
+
+See the "crontab_conversion.go" file in the previous section.
+
+**Step 3c**: Update the CRD to add the second version.
+
+Do this by adding a new item to the "versions" list, containing the new data model:
+
+|crd2.yaml|
+|-|
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: crontabs.example.com
+spec:
+  group: example.com
+  versions:
+  - name: v1
+    served: true
+    storage: true
+    schema:
+      properties:
+        spec:
+          properties:
+            cronSpec:
+              type: string
+            image:
+              type: string
+  - name: v2
+    served: true
+    storage: false
+    schema:
+      properties:
+        spec:
+          properties:
+            min:
+              type: string
+            hour:
+              type: string
+            dayOfMonth:
+              type: string
+            month:
+              type: string
+            dayOfWeek:
+              type: string
+            image:
+              type: string
+  scope: Namespaced
+  names:
+    plural: crontabs
+    singular: crontab
+    kind: CronTab
+    shortNames:
+    - ct
+  conversion:
+    strategy: external
+    webhook:
+      client_config:
+        namespace: crontab
+        service: crontab_conversion
+        Path: /crontab_convert
+```
+
+
+And apply it:
+
+```bash
+Kubectl apply -f crd2.yaml
+```
+
+
+**Step 5**: add a new CR in v2:
+
+|cr2.yaml|
+|-|
+```yaml
+apiVersion: "stable.example.com/v2"
+kind: CronTab
+metadata:
+  name: my-second-cron-object
+spec:
+  min: "*"
+  hour: "*"
+  day_of_month: "*"
+  dayOfWeek: "*/5"
+  month: "*"
+  image: my-awesome-cron-image
+```
+
+
+And create it:
+
+```bash
+Kubectl create -f cr2.yaml
+```
+
+
+**Step 6**: storage now has two custom resources in two different versions. To downgrade to previous CRD, one can apply crd1.yaml but that will fail as the status.storedVersions has both v1 and v2 and those cannot be removed from the spec.versions list. To downgrade, first create a crd2-b.yaml file that sets v1 as storage version and apply it, then follow "*Upgrade existing objects to a new stored version*“ in [this document](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/). After all CRs in the storage has v1 version, you can apply crd1.yaml.
```

Comment:
Why does the storage have two versions? I see v1 is still the storage version. We only have v1 in etcd.